Towny/src/main/java/com/palmergames/bukkit/towny/command/TownCommand.java

```diff
@@ -4459,4 +4479,85 @@ private static int homeBlockDistance(Town town1, Town town2) {
 
 		return (int) MathUtil.distance(town1.getHomeBlockOrNull().getCoord(), town2.getHomeBlockOrNull().getCoord());
 	}
+
+	private void parseTownForSaleCommand(Player player, String[] args) throws TownyException {
+		if (args.length == 0)
+			throw new TownyException(Translatable.of("msg_error_must_be_num"));
+
+		double forSalePrice = Double.parseDouble(args[0]);
+
+		Town town = getTownFromPlayerOrThrow(player);
+
+		Confirmation
+			.runOnAccept(() -> {
+				setTownForSale(town, forSalePrice, false);
+				TownyMessaging.sendPrefixedTownMessage(town, Translatable.of("msg_town_forsale", town.getName(), TownyEconomyHandler.getFormattedBalance(forSalePrice)));
+			})
+			.setTitle(Translatable.of("msg_town_sell_confirmation", TownyEconomyHandler.getFormattedBalance(forSalePrice)))
+			.sendTo(player);
+	}
+
+	private void parseTownNotForSaleCommand(Player player) throws TownyException {
+		Town town = getTownFromPlayerOrThrow(player);
+
+		if (town.isForSale()) {
+			setTownNotForSale(town, false);
+			TownyMessaging.sendPrefixedTownMessage(town, Translatable.of("msg_town_notforsale", town.getName()));
+		}
+	}
+
+	public static void setTownForSale(Town town, double price, boolean admin) {
+		if (town != null) {
+			town.setForSale(true);
+			town.setForSalePrice(price);
+			town.save();
+		}
+	}
+
+	public static void setTownNotForSale(Town town, boolean admin) {
+		if (town != null) {
+			town.setForSale(false);
+			town.save();
+		}
+	}
+
+	private void parseTownBuyTownCommand(Player player, String[] args) throws TownyException {
+		catchConsole(player);
+
+		if (args.length == 0) {
+			TownyMessaging.sendErrorMsg(player, Translatable.of("msg_specify_name"));
+			return;
+		}
+
+		Town town = TownyUniverse.getInstance().getTown(args[0]);
+		if (town == null) {
+			throw new TownyException(Translatable.of("msg_err_town_unknown", args[0]));
+		}
+
+		if (!town.isForSale()) {
+			throw new TownyException(Translatable.of("msg_town_buytown_not_forsale"));
+		}
+
+		Resident resident = getResidentOrThrow(player);
+		if (town.getMayor() == resident) {
+			throw new TownyException(Translatable.of("msg_town_buytown_already_mayor", town.getName()));
+		}

```

Comment:
```suggestion
		if (resident.isMayor()) {
			throw new TownyException(Translatable.of("msg_town_buytown_already_mayor", resident.getTownOrNull().getName()));
		}

```

I think it ought to stop any mayor from buying a town, as any mayor that buys a town will stop being a mayor of their own town.

This suggestion above is not as good as having something more descriptive when `town != resident.getTownOrNull()`, which I think should explain that the player has to set a new mayor before they can buy another town. Similar to when a mayor does `/t leave`.